### PR TITLE
dataTrack support added.

### PR DESCRIPTION
**Pull request by UberMC**
dataTracks for using Twilios 'UDP like' networking to send packets between clients.

onDataTrackMessageReceived, 
sendString, 
onParticipantAddedDataTrack, 
onParticipantRemovedDataTrack

iOS and android tested in a very complex activity based video call app.
No crashes, No errors on logcat, react native, gradle, xCode.

The only things I decided not to fully complete, and I don't really wish to:
- onParticipantAddedDataTrack/ onParticipantRemovedDataTrack the events don't have the most useful information, I didn't need that information.
- the dataTrack isn't part of the 'stats' report.
- sendString is only supported I didn't add sendData. As in react native it's very easy to stringify/parse a JSON object. Though maybe sendData is better performance I don't know.

### CODE REVIEW

File: ios/RCTTWVideoModule.m [hunk 1]
```diff
@@ -434,6 +466,17 @@ - (void)remoteParticipant:(TVIRemoteParticipant *)participant disabledAudioTrack
...
 }
 
+# pragma mark - TVIRemoteDataTrackDelegate
+
+- (void)remoteDataTrack:(nonnull TVIRemoteDataTrack *)remoteDataTrack didReceiveString:(nonnull NSString *)message {
+    [self sendEventCheckingListenerWithName:dataTrackMessageReceived body:@{ @"message": message }];
+}
+
+- (void)remoteDataTrack:(nonnull TVIRemoteDataTrack *)remoteDataTrack didReceiveData:(nonnull NSData *)message {
+}
```
Comment (slycoder) [inline ios/RCTTWVideoModule.m]: Maybe add a TODO or a NSLog here?

File: ios/RCTTWVideoModule.m [hunk 2]
```diff
@@ -361,9 +383,10 @@ - (void)didConnectToRoom:(TVIRoom *)room {
     [participants addObject:[p toJSON]];
   }
   TVILocalParticipant *localParticipant = room.localParticipant;
+  //localParticipant.publishDataTrack(localDataTrack)
```
Comment (slycoder) [inline ios/RCTTWVideoModule.m]: Remove

File: ios/RCTTWVideoModule.m [hunk 3]
```diff
@@ -311,12 +318,27 @@ -(NSMutableDictionary*)convertLocalVideoTrackStats:(TVILocalVideoTrackStats *)st
...
+    }
+
     builder.roomName = roomName;
   }];
 
   self.room = [TwilioVideo connectWithOptions:connectOptions delegate:self];
 }
 
+RCT_EXPORT_METHOD(sendString:(nonnull NSString *)message) {
+    NSLog(@"Send String!");
```
Comment (slycoder) [inline ios/RCTTWVideoModule.m]: Some debugging stuff could be removed =).

File: src/TwilioVideo.ios.js [hunk 4]
```diff
@@ -303,6 +340,11 @@ export default class extends Component {
           this.props.onParticipantDisabledAudioTrack(data)
         }
       }),
+      this._eventEmitter.addListener('dataTrackMessageReceived', data => {
+        if (this.props.onDataTrackMessageReceived) {
+          this.props.onDataTrackMessageReceived(data)
```
Comment (slycoder) [inline src/TwilioVideo.ios.js]: Does this need to unwrap it from the message field of  data?

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java [hunk 5]
```diff
@@ -869,6 +949,17 @@ private WritableMap buildParticipant(Participant participant) {
         return participantMap;
     }
 
+
+    private WritableMap buildParticipantDataEvent(Participant participant) {
+        WritableMap participantMap = buildParticipant(participant);
+        WritableMap participantMap2 = buildParticipant(participant);
```
Comment (slycoder) [inline android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java]: Shouldn't these be the fields for the track?

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java [hunk 6]
```diff
@@ -178,6 +205,15 @@ public CustomTwilioVideoView(ThemedReactContext context) {
         audioManager = (AudioManager) themedReactContext.getSystemService(Context.AUDIO_SERVICE);
         myNoisyAudioStreamReceiver = new BecomingNoisyReceiver();
         intentFilter = new IntentFilter(Intent.ACTION_HEADSET_PLUG);
+
+        // Create the local data track
+       // localDataTrack = LocalDataTrack.create(this);
+       localDataTrack = LocalDataTrack.create(getContext());
+
+       // Start the thread where data messages are received
+        dataTrackMessageThread.start();
```
Comment (slycoder) [inline android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java]: Is there any cleanup that needs to happen with this thread?

File: ios/RCTTWVideoModule.m (hunk 1 above)
Comment (UberMC) [reply ios/RCTTWVideoModule.m]: done

File: ios/RCTTWVideoModule.m (hunk 2 above)
Comment (UberMC) [reply ios/RCTTWVideoModule.m]: done

File: ios/RCTTWVideoModule.m (hunk 3 above)
Comment (UberMC) [reply ios/RCTTWVideoModule.m]: done

File: src/TwilioVideo.ios.js (hunk 4 above)
Comment (UberMC) [reply src/TwilioVideo.ios.js]: Looks good to me

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java (hunk 5 above)
Comment (UberMC) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java]: Perhaps you can help here? Not sure how to pass this kind of track and output the relevant information important to this type of track.

I'm sorry this event information just isn't important to me right now.

File: android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java (hunk 6 above)
Comment (UberMC) [reply android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java]: Done onHostDestroy dataTrackMessageThread.quit();

**Comment by UberMC** [comment]
Hey thanks for the feedback. I made the requested changes except for buildParticipantDataEvent logging the track properties as I just have no use for that right now

This has been tested thoroughly and I will be moving on from this until a breaking change occurs. Thanks bye.

**Comment by slycoder** [comment]
Merging soon, and I'll fix the lint errors in a followup PR.

**Pull request merged**